### PR TITLE
More mod info

### PR DESCRIPTION
**Pull request by Geometrically**
More mod info, may also add icon/ user editing routes

### CODE REVIEW

File: migrations/20201122043349_more-mod-data.sql
Comment (Aeledfyr) [inline migrations/20201122043349_more-mod-data.sql:44]: This should not have a default, or licenses should have an unspecified variant.  This should be required to be specified in the future, but it's not safe to assume that all current mods are all rights reserved.

File: migrations/20201122043349_more-mod-data.sql
Comment (Aeledfyr) [inline migrations/20201122043349_more-mod-data.sql:48]: It would also be good to have an unknown variant of this as the default for old mods, that can't be selected when creating a new mod.  That way we know what information is valid and what we have to fix.

File: src/database/models/mod.rs
Comment (Aeledfyr) [inline src/database/models/mod.rs:107]: This should be `&str`

File: src/database/models/mod.rs
Comment (Aeledfyr) [inline src/database/models/mod.rs:129]: This should be `&str`

File: src/models/mods.rs
Comment (Aeledfyr) [inline src/models/mods.rs:79]: This could be reduced down to `write!(fmt, "{}", self.as_str())`

File: src/routes/mod_creation.rs
Comment (Aeledfyr) [inline src/routes/mod_creation.rs:129]: If licenses are strictly limited to the options we give, there needs to be an "other" category and a way of linking to the license.  People will always come up with weird licenses, so we can't enumerate them all.

File: src/routes/mod_creation.rs
Comment (Aeledfyr) [inline src/routes/mod_creation.rs:545]: This shouldn't be empty; can you query this when you get the id?

File: src/routes/mods.rs
Comment (Aeledfyr) [inline src/routes/mods.rs:133]: Make sure this gets finished before merging

**Review by Aeledfyr — CHANGES_REQUESTED** [review]
For the licenses, how do you plan on supporting  the large variety of licenses people will use?  How do you plan on supporting dual licensed mods?  We also have to have a way of supporting crayon (custom) licenses.

There's also the question of how to deal with the defaults for mods currently in the system.  We should have an unknown / unspecified variant for each (license and client-/server-side) that is used as the default for old mods, but require new mods to specify a real value.

File: migrations/20201122043349_more-mod-data.sql
Comment (Geometrically) [reply migrations/20201122043349_more-mod-data.sql:48]: Required is best to be the default, as unknown gives users not much information, while it is safe to assume that every mod is required on the server and client by default.

Comment (Geometrically) [reply migrations/20201122043349_more-mod-data.sql:48]: And the default is only for the data currently in the database

File: migrations/20201122043349_more-mod-data.sql
Comment (Geometrically) [reply migrations/20201122043349_more-mod-data.sql:44]: It is safe to assume that all mods are ARR, unless specified. It also saves us from legal trouble, because if we specify the wrong license for a mod (on our end), that would be bad for us

File: src/routes/mod_creation.rs
Comment (Geometrically) [reply src/routes/mod_creation.rs:129]: I'll add a "Custom License" license in the migration for the sake of this

File: migrations/20201122043349_more-mod-data.sql
Comment (Aeledfyr) [inline migrations/20201122043349_more-mod-data.sql:62]: This is not anonymized, and should probably be deleted after a few days.  Most users will be using IPv4 (Do we even support IPv6? We should), so that gives a bit less than 2^32 addresses - this ends up taking <30 minutes to generate the hash of *every* IPv4 address, allowing this to be entirely de-anonymized.

I'm assuming that this is intended to avoid tracking multiple downloads from the same IP in a short time range?  If so, I  would recommend clearing this field for every record older than an hour.  It would also be good to generate a random salt on startup that is hashed with the IP address, and ensure that that salt is never logged or saved, but that doesn't entirely fix the issue; we need to clear old records on a regular basis.

File: src/routes/versions.rs
Comment (Aeledfyr) [inline src/routes/versions.rs:590]: Why is this a post request?  Shouldn't it be get?

File: src/routes/moderation.rs
Comment (Aeledfyr) [inline src/routes/moderation.rs:62]: This needs to have filled out data for moderation to be usable. This should use the same system as the normal mods route if possible - it will be slower, but you should be able to query a list of ids and then use the exact same code as used in the mods route, so you end up with the full information here.

File: src/routes/versions.rs
Comment (Aeledfyr) [inline src/routes/versions.rs:616]: This checks whether the mod has been downloaded in the last 30 minutes by *anyone*, not from this specific ip.

File: src/routes/versions.rs
Comment (Aeledfyr) [inline src/routes/versions.rs:639]: Isn't this supposed to be the sha-1 hash of the IP address?  Also, this should not throw any errors if it can't log it - it should still redirect and let the user download it.

**Review by Aeledfyr — CHANGES_REQUESTED** [review]
There is still some information not being returned by the moderation / other routes (also see previous review comments).  The download tracking system has issues, and isn't great from a privacy standpoint.  The best solution for the privacy issues with the download counting would be to either clear the ip field or just delete the download records for downloads that are more than a few hours old; in addition, hash the ips with a random salt generated at startup.

File: migrations/20201122043349_more-mod-data.sql
Comment (diabolical17) [reply migrations/20201122043349_more-mod-data.sql:62]: Commenting on the IPv6, its becoming slowly standardised, i suggest using it instead of having to rewrite code in the future. You could use IPv4 as a fallback if IPv6 is unavailable or something 🤷🏽

Comment (Geometrically) [reply migrations/20201122043349_more-mod-data.sql:62]: `realip_remote_addr` supports IPV6 iirc

Comment (diabolical17) [reply migrations/20201122043349_more-mod-data.sql:62]: Ah, nevermind then 😅

File: src/routes/moderation.rs
Comment (Geometrically) [reply src/routes/moderation.rs:62]: This isn't needed, as this is just a list of mods that the moderator needs to approve. When reviewing a mod, a mod will then click on the mods on the frontend and go to the mod's page, which will have this data.

File: migrations/20201122043349_more-mod-data.sql
Comment (Aeledfyr) [reply migrations/20201122043349_more-mod-data.sql:62]: The IPv6 comment is the fact that our current server hosting stuff doesn't support IPv6.  This is either an issue with the DNS records or the actual host configuration (@Geometrically?)

Comment (Geometrically) [reply migrations/20201122043349_more-mod-data.sql:62]: Wdym? The server isn't on IPV6, but the host should accept IPV6 requests

File: src/routes/moderation.rs
Comment (Aeledfyr) [reply src/routes/moderation.rs:62]: How the *frontend* is structured shouldn't affect the backend design.  If this isn't going to be included, make a new structure for it that doesn't include the data that isn't included.  This route should either return the *exact* same information that the normal mods route does, or return *only* a list of mod ids that need to be reviewed, which would then require sending it to the mods route to get the actual information.

This either has to return the full data, or make a new partial mod data structure that specifies what information is actually being sent instead of filling it in with defaults.

File: .env
Comment (Aeledfyr) [inline .env:39]: This isn't what I was talking about; this needs to be generated randomly on server startup, or it isn't useful at all.  If it's static, even if it's configured to be something else on the production host, then it is just as easy to brute force as without it.

This isn't as much of an issue with older downloads being cleared, but this doesn't do much as is - it should be easy to add a random number to the end on startup and then pass that to the routes using actix's shared data.

File: src/database/models/categories.rs
Comment (Aeledfyr) [inline src/database/models/categories.rs:567]: Copy-pasted docs, go through and fix the other copies of this - also in the insert method.

**Review by Aeledfyr — CHANGES_REQUESTED** [review]
A few things - the moderation routes need to return the same data as the normal mod routes, and the salt currently isn't implemented properly.  (Also, the build is failing.  Did you run `cargo sqlx prepare`?)

**Review by Aeledfyr — APPROVED** [review]
Don't merge this until the frontend supports the changes to mod creation, but it looks like it should be ready to go.

**Pull request merged**
